File: src/main/java/com/bleudev/nine_lifes/Nine_lifes.java
package com.bleudev.nine_lifes;

import com.bleudev.nine_lifes.custom.*;
import com.bleudev.nine_lifes.interfaces.mixin.LivingEntityCustomInterface;
import com.bleudev.nine_lifes.networking.Packets;
import com.bleudev.nine_lifes.networking.payloads.JoinMessage;
import com.bleudev.nine_lifes.networking.payloads.StartChargeScreenEffect;
import com.bleudev.nine_lifes.networking.payloads.UpdateCenterHeart;
import com.bleudev.nine_lifes.util.ComponentUtils;
import com.bleudev.nine_lifes.util.LivesUtils;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback;
import net.fabricmc.fabric.api.entity.event.v1.ServerPlayerEvents;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.fabricmc.fabric.api.registry.FabricBrewingRecipeRegistryBuilder;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.LivingEntity;
import net.minecraft.item.Items;
import net.minecraft.potion.Potions;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.math.Box;
import net.minecraft.world.World;
import net.minecraft.world.explosion.EntityExplosionBehavior;

import static com.bleudev.nine_lifes.compat.VersionCompat.getPosCompat;
import static com.bleudev.nine_lifes.util.ComponentUtils.item_ensure_custom_foods;
import static com.bleudev.nine_lifes.util.ComponentUtils.should_update_amethyst_shard;

public class Nine_lifes implements ModInitializer {
    public static final String MOD_ID = "nine_lifes";
    public static final String NAME = "Nine lifes";
    public static final String AUTHOR = "bleudev";
    public static final String VERSION = "1.8";
    public static final String GITHUB_LINK = "https://github.com/bleudev/nine_lifes";
    public static final String MODRINTH_LINK = "https://modrinth.com/mod/nine_lifes";

    @Override
    public void onInitialize() {
        Packets.initialize();
        CustomEffects.initialize();
        CustomConsumeEffectTypes.initialize();
        CustomEnchantments.initialize();
        CustomPotions.initialize();
        CustomEntities.initialize();
        FabricBrewingRecipeRegistryBuilder.BUILD.register(builder -> {
            builder.registerPotionRecipe(Potions.WATER, Items.AMETHYST_SHARD, CustomPotions.AMETHYSM);
        });
        ServerPlayerEvents.JOIN.register(player -> {
            if ((!player.isSpectator()) && (LivesUtils.getLives(player) == 0))
                LivesUtils.resetLives(player);
            ServerPlayNetworking.send(player, new UpdateCenterHeart(LivesUtils.getLives(player)));
            ServerPlayNetworking.send(player, new JoinMessage(LivesUtils.getLives(player)));
        });
        ServerTickEvents.END_SERVER_TICK.register(server -> {
            // Custom food (amethyst shard)
            for (var player: server.getPlayerManager().getPlayerList())
                ComponentUtils.player_ensure_custom_foods(player);

            // Find all amethyst shard item entities in `amethyst_shard_find_radius`
            double amethyst_shard_find_radius = 20;
            for (var world: server.getWorlds()) {
                for (var player: world.getPlayers()) {
                    var box = Box.of(getPosCompat(player), amethyst_shard_find_radius, amethyst_shard_find_radius, amethyst_shard_find_radius);
                    world.getEntitiesByType(
                        EntityType.ITEM,
                        box,
                        entity -> entity.getStack().isOf(Items.AMETHYST_SHARD) &&
                                         should_update_amethyst_shard(entity.getStack())
                    ).forEach(entity -> entity.setStack(item_ensure_custom_foods(entity.getStack())));
                    world.getEntitiesByClass(LivingEntity.class, box, ignored -> true).forEach(entity -> {
                        var inter = (LivingEntityCustomInterface) entity;
                        int damage_ticks = inter.nine_lifes$getDamageTicks();
                        if (damage_ticks == -1) return;

                        if (damage_ticks == 0) {
                            world.createExplosion(
                                    entity, CustomDamageTypes.of(world, CustomDamageTypes.CHARGED_AMETHYST_DAMAGE_TYPE), new EntityExplosionBehavior(entity),
                                    entity.getX(), entity.getY(), entity.getZ(),
                                    7f, true, World.ExplosionSourceType.MOB
                            );
                            entity.damage(world,
                                CustomDamageTypes.of(entity, CustomDamageTypes.CHARGED_AMETHYST_DAMAGE_TYPE),
                                1000f
                            );
                        }
                        inter.nine_lifes$setDamageTicks(--damage_ticks);
                    });
                }
                tryChargeItems(world);
            }

            // Wind charges
            for (var world: server.getWorlds())
                world.getEntitiesByType(EntityType.WIND_CHARGE, ignored -> true).forEach(wind_charge -> {
                    WindChargeTickFeatures.do_for(world, wind_charge);
                });
        });
        CommandRegistrationCallback.EVENT.register((dispatcher, registryAccess, environment) -> {
            CustomCommands.initialize(dispatcher);
        });
    }

    private void tryChargeItems(ServerWorld world) {
        var lightning_charging_radius = 1;
        var charge_screen_effect_radius_min = 3;
        var charge_screen_effect_radius_max = 20;
        var charge_screen_effect_radius_diff = charge_screen_effect_radius_max - charge_screen_effect_radius_min;
        var charge_screen_max_strength = 0.5;
        var charge_screen_duration = 6;

        final var CHARGE_ENCHANTMENT = CustomEnchantments.getEntry(world.getRegistryManager(), CustomEnchantments.CHARGE);
        world.getEntitiesByType(EntityType.LIGHTNING_BOLT, ignored -> true).forEach(lightning -> {
            world.getEntitiesByType(EntityType.ITEM,
                    Box.of(getPosCompat(lightning), lightning_charging_radius, lightning_charging_radius, lightning_charging_radius),
                    entity -> entity.getStack().isIn(CustomTags.ItemTags.LIGHTNING_CHARGEABLE)
            ).forEach(item_entity -> {
                var stack = item_entity.getStack();
                if (!stack.getEnchantments().getEnchantments().contains(CHARGE_ENCHANTMENT)) {
                    stack.addEnchantment(CHARGE_ENCHANTMENT, 1);
                    item_entity.setStack(stack);

                    world.getPlayers().forEach(player -> {
                        var distance = Math.max(getPosCompat(player).distanceTo(getPosCompat(item_entity)) - charge_screen_effect_radius_min, 0);
                        var strength = charge_screen_max_strength * (charge_screen_effect_radius_diff - distance) / charge_screen_effect_radius_diff;
                        ServerPlayNetworking.send(player, new StartChargeScreenEffect(charge_screen_duration, (float) strength));
                    });
                }
            });
        });
    }
}

File: 1.21.6/src/main/kotlin/com/bleudev/nine_lifes/NineLifes.kt
package com.bleudev.nine_lifes

import com.bleudev.nine_lifes.api.event.EntityLifecycleEvents
import com.bleudev.nine_lifes.custom.*
import com.bleudev.nine_lifes.custom.NineLifesEntities.WANDERING_ARMOR_STAND
import com.bleudev.nine_lifes.custom.packet.payload.*
import com.bleudev.nine_lifes.util.*
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.entity.event.v1.ServerLivingEntityEvents
import net.fabricmc.fabric.api.entity.event.v1.ServerPlayerEvents
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerTickEvents
import net.fabricmc.fabric.api.registry.FabricBrewingRecipeRegistryBuilder
import net.minecraft.core.component.DataComponents
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.EntitySpawnReason
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.entity.projectile.windcharge.WindCharge
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.alchemy.PotionContents
import net.minecraft.world.item.alchemy.Potions
import net.minecraft.world.item.crafting.Ingredient
import net.minecraft.world.level.ExplosionDamageCalculator
import net.minecraft.world.level.GameType
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BrewingStandBlockEntity
import net.minecraft.world.level.entity.EntityTypeTest

class NineLifes : ModInitializer {
    override fun onInitialize() {
        NineLifesPackets.initialize()
        NineLifesMobEffects.initialize()
        NineLifesConsumeEffects.initialize()
        NineLifesEnchantments.initialize()
        NineLifesPotions.initialize()
        NineLifesEntities.initialize()
        NineLifesCommands.initialize()
        FabricBrewingRecipeRegistryBuilder.BUILD.register { it.registerPotionRecipe(
            Potions.WATER,
            Ingredient.of(Items.AMETHYST_SHARD),
            NineLifesPotions.AMETHYSM
        ) }
        ServerPlayerEvents.JOIN.register { player ->
            if ((!player.isSpectator) && player.lifes == 0) player.resetLifes()
            val lifes = player.lifes
            player.sendPackets(UpdateLifesCount(lifes), JoinMessage(lifes))
            if (isInBetaMode()) player.sendPacket(BetaModeMessage.INSTANCE)
        }
        ServerPlayerEvents.AFTER_RESPAWN.register { _, newPlayer, _ ->
            newPlayer.sendPacket(AfterPlayerRespawn.INSTANCE)
        }
        ServerTickEvents.END_SERVER_TICK.register { server ->
            for (player in server.playerList.players) playerEnsureCustomFoods(player)

            for (level in server.allLevels) {
                for (player in level.players()) {
                    val box = ofDXYZ(player.position(), 20)
                    level.getEntities(EntityType.ITEM) { entity ->
                        entityIn<ItemEntity>(box)(entity) &&
                        entity.item.`is`(Items.AMETHYST_SHARD) &&
                        shouldUpdateAmethystShard(entity.item)
                    }.forEach { entity -> entity.item = itemEnsureCustomFoods(entity.item) }
                    level.getEntities(EntityTypeTest.forClass(LivingEntity::class.java), entityIn(box)).forEach { entity ->
                        when (entity.damageTicks) {
                            -1 -> return@forEach
                            0 -> {
                                level.explode(entity, NineLifesDamageSources.charged(level),
                                    ExplosionDamageCalculator(), entity.position(),
                                    7f, true, Level.ExplosionInteraction.MOB)
                                entity.hurtServer(level, NineLifesDamageSources.charged(level), Float.MAX_VALUE)
                            }
                        }
                        entity.damageTicks -= 1
                    }
                }
                tryChargeItems(level)
            }

            for (world in server.allLevels) world.getEntities(EntityType.WIND_CHARGE, alwaysTrue())
                .forEach { tryWindChargeFeatures(world, it) }
        }
        EntityLifecycleEvents.ENTITY_SPAWN.register { entity, level ->
            if (entity.type.equals(EntityType.ARMOR_STAND)) {
                if (level.getRandom().nextFloat() < WANDERING_ARMOR_STAND_SPAWN_CHANCE) {
                    val newEntity = WANDERING_ARMOR_STAND.create(level, EntitySpawnReason.SPAWN_ITEM_USE)
                    if (newEntity != null) {
                        newEntity.setPos(entity.position())
                        newEntity.copyPosition(entity)
                        level.addFreshEntity(newEntity)
                        entity.discard()
                    }
                }
            }
        }
        ServerLivingEntityEvents.AFTER_DEATH.register { entity, damageSource ->
            if (entity is ServerPlayer && entity.gameMode().isSurvival) {
                entity.addLifes(if (damageSource.`is`(NineLifesDamageTypeTags.GIVES_LIFE)) 1 else -1)
                if (entity.lifes <= 0) entity.setGameMode(GameType.SPECTATOR)
            } }
    }

    private fun tryChargeItems(level: ServerLevel) {
        val chargeScreenEffectRadiusDiff = CHARGE_SCREEN_EFFECT_RADIUS_MAX - CHARGE_SCREEN_EFFECT_RADIUS_MIN

        val chargeEnchantment = NineLifesEnchantments.Holders.charge(level.registryAccess())
        level.getEntities(EntityType.LIGHTNING_BOLT, alwaysTrue()).forEach { lightning ->
            level.getEntities(
                EntityType.ITEM,
                entityIn<ItemEntity>(lightning.position(), LIGHTNING_CHARGING_RADIUS)
                .and({ entity -> entity.item.`is`(NineLifesItemTags.LIGHTNING_CHARGEABLE) }))
                .forEach { itemEntity ->
                    if (itemEntity.item.enchantments.getLevel(chargeEnchantment) == 0) {
                        itemEntity.item.enchant(chargeEnchantment, 1)
                        level.players().forEach { player ->
                            val distance = (player.position().distanceTo(itemEntity.position()) - CHARGE_SCREEN_EFFECT_RADIUS_MIN)
                                .coerceAtLeast(.0)
                            val strength = CHARGE_SCREEN_MAX_STRENGTH * (chargeScreenEffectRadiusDiff - distance) / chargeScreenEffectRadiusDiff
                            player.sendPacket(StartChargeScreen(CHARGE_SCREEN_DURATION, strength.toFloat()))
                        }
                    }
                }
        }
    }

    fun tryWindChargeFeatures(level: ServerLevel, windCharge: WindCharge) {
        val actionBox = ofDXYZ(windCharge.position(), 3)
        forBlocksInBox(actionBox) { pos ->
            val blockEntity = level.getBlockEntity(pos)
            if (blockEntity is BrewingStandBlockEntity) {
                if (blockEntity.items?.subList(0, 3)?.stream()?.noneMatch { potion ->
                    try {
                        for (eff in potion.get(DataComponents.POTION_CONTENTS)?.allEffects ?: listOf())
                            if (eff.effect.equals(NineLifesMobEffects.AMETHYSM)) return@noneMatch true
                        return@noneMatch false
                    } catch (_: NullPointerException) {
                        return@noneMatch false
                    }
                } ?: true) return@forBlocksInBox 0

                level.removeBlock(pos, false)
                level.explode(
                    null, NineLifesDamageSources.charged(level), null,
                    pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble(),
                    3f, true, Level.ExplosionInteraction.BLOCK
                )
            }
            0
        }
        level.getEntities(EntityTypeTest.forClass(LivingEntity::class.java), entityIn(actionBox)).forEach {
            it.removeEffect(NineLifesMobEffects.AMETHYSM)
        }

        level.players().forEach { player ->
            if (!actionBox.contains(player.position())) return@forEach
            val inventory = player.getInventory()
            var inventoryUpdated = false
            for (slot in 0..<inventory.toList().size) {
                val stack = inventory.getItem(slot)
                var newStack = stack.copy()

                val potion: PotionContents?
                if ((stack.get(DataComponents.POTION_CONTENTS).also { potion = it }) != null)
                    for (effect in potion?.allEffects ?: listOf()) if (effect.effect.equals(NineLifesMobEffects.AMETHYSM)) {
                    newStack = ItemStack.EMPTY
                    break
                }

                if (!ItemStack.isSameItem(stack, newStack)) {
                    inventory.setItem(slot, newStack)
                    inventoryUpdated = true
                }
            }

            // TODO: Broke sound
        }
    }
}